File: PragDaveKatas/Elixir/06_anagrams/anagrams.exs
#!/usr/bin/env elixir

parse_string  = &String.strip/1
word_and_idx  = &({&1, to_char_list(&1) |> :lists.sort |> to_string })
anagram_idx   = &(elem(&1,1))
anagram_word  = &(elem(&1,0))
sort_idx      = &(anagram_idx.(&1) < anagram_idx.(&2))
is_anagram?   = &(Enum.at(&1,1) != nil)
extract_word  = &(Enum.map(&1, anagram_word))

anagram_list = File.stream!("wordlist.txt")
            |> Stream.map(parse_string)
            |> Enum.map(word_and_idx)
            |> Enum.sort(sort_idx)
            |> Enum.chunk_by(anagram_idx)
            |> Enum.map(extract_word)
            |> Enum.filter(is_anagram?)

IO.puts "Found #{length(anagram_list)} anagrams"

find_longest = fn x,a when length(x) > length(a) -> x
                  _,a                            -> a
               end

longest = Enum.reduce(anagram_list, find_longest)
IO.puts "Longest anagram list is #{inspect longest} with #{inspect length(longest)} words"

File: PragDaveKatas/Python/06_anagrams/anagrams.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-

def word_idx(word): return "".join(sorted(word))

def read_words(filename):
    with open(filename) as filehandle:
        words    = (line.strip() for line in filehandle)
        wordlist = [( word, word_idx(word)) for word in words]
    return wordlist

def chunk(words):
    ret, idx = [], ""
    for w in words:
        if w[1] != idx:
            if ret != []: yield ret
            ret, idx = [w[0]], w[1]
        else:
            ret.append(w[0])
    if ret != []: yield ret
    return

def find_longest(acc, x): return x if len(x) > len(acc) else acc

if __name__ == "__main__":
    wordlist = read_words("wordlist.txt")
    wordlist = sorted(wordlist, key=lambda x: x[1])
    anagramlist = chunk(wordlist)
    anagramlist = filter(lambda x: len(x) > 1, anagramlist)
    print "Found", len(anagramlist), "anagrams"
    longest = reduce(find_longest, anagramlist)
    print "Longest anagram list is", longest, "with", len(longest), "anagrams"

